### PR TITLE
JWT auth defaults

### PR DESCRIPTION
**Pull request by shamsimam**
## Changes proposed in this PR

- adds option to enable JWT auth by default for tokens
- adds JWT auth support for waiter api requests

## Why are we making these changes?

We would like to allow Waiter API requests and service requests to get JWT authentication by default.

### CODE REVIEW

File: waiter/src/waiter/auth/jwt.clj [hunk 1]
```diff
@@ -365,12 +365,19 @@
...
-             (auth/select-auth-header request access-token?))
+             (auth/select-auth-header request access-token?)
+             (or
+               ;; service requests will enable JWT auth based on env variable or when absent, the use-bearer-auth-default?
+               (and (not waiter-api-call?)
+                    (= "true" (get-in request [:waiter-discovery :service-parameter-template "env" "USE_BEARER_AUTH"]
+                                      (str use-bearer-auth-default?))))
+               ;; waiter api requests will enable JWT auth based on use-bearer-auth-default?
+               (and waiter-api-call?
+                    use-bearer-auth-default?)))
```
Comment (shamsimam) [inline waiter/src/waiter/auth/jwt.clj:380]: Do you want a separate flag to configure the defaults for waiter api requests?

File: waiter/config-full.edn [hunk 2]
```diff
@@ -248,7 +248,15 @@
                               ;; The token type expected in the access token header
                               :token-type "JWT"
                               ;; The interval at which to refresh the keys from the authorization server:
-                              :update-interval-ms 60000}
+                              :update-interval-ms 60000
+                              ;; Whether JWT authentication is the default for Waiter API requests.
+                              ;; For services, the value of USE_BEARER_AUTH environment variable overrides this value.
+                              ;; The default is true.
+                              :use-bearer-auth-default-for-api? true
```
Comment (sradack) [inline waiter/config-full.edn:255]: Would this be clearer as `:allow-bearer-auth-default-for-api?`.  `use` implies other methods aren't allowed.

Comment (sradack) [reply waiter/config-full.edn:255]: What does `default` mean here?  Should this just be `:allow-bearer-auth-api?`

File: waiter/config-full.edn [hunk 3]
```diff
@@ -248,7 +248,15 @@
...
+                              ;; Whether JWT authentication is the default for services proxy requests.
+                              ;; The value of USE_BEARER_AUTH environment variable of the service overrides this value.
+                              ;; The default is false.
+                              :use-bearer-auth-default-for-service? false}
```
Comment (sradack) [inline waiter/config-full.edn:259]: Would this be clearer as :allow-bearer-auth-default-for-service?. `use` implies other methods aren't allowed.

Comment (sradack) [reply waiter/config-full.edn:259]: What does `default` mean here?  Should this just be `:allow-bearer-auth-services?`

File: waiter/config-full.edn (hunk 2 above)
Comment (shamsimam) [reply waiter/config-full.edn]: Renamed to `:allow-bearer-auth-api?`. Rephrased the docstring, this is an optional field, the default refers to the value used when the field is not configured.

File: waiter/config-full.edn (hunk 3 above)
Comment (shamsimam) [reply waiter/config-full.edn]: Renamed to `:allow-bearer-auth-services?`. Rephrased the docstring, this is an optional field, the default refers to the value used when the field is not configured.

File: waiter/src/waiter/auth/jwt.clj (hunk 1 above)
Comment (shamsimam) [reply waiter/src/waiter/auth/jwt.clj]: Added separate flag.

**Pull request merged**